catalog/src/index.ts

```diff
@@ -0,0 +1,200 @@
+import type { Api } from "@codemod.com/workflow";
+import * as semver from "semver";
+
+type PackagesVersions = Record<
+	string,
+	{
+		usages: number;
+		versions: string[];
+	}
+>;
+
+type PackageJson = {
+	packageManager?: string;
+	dependencies?: Record<string, string>;
+	devDependencies?: Record<string, string>;
+};
+
+const isAlias = (version: string) => {
+	return version.match(/^p?npm:/)?.length === 1;
+};
+
+const validRange = (version: string) => {
+	return semver.validRange(version) !== null;
+};
+
+const readDependencies = (
+	packagesVersions: PackagesVersions,
+	dependencies: Record<string, string> = {},
+) => {
+	for (const [name, version] of Object.entries(dependencies)) {
+		if (
+			version === "workspace:*" ||
+			(!isAlias(version) && !validRange(version))
+		) {
+			continue;
+		}
+
+		if (!packagesVersions[name]) {
+			packagesVersions[name] = {
+				usages: 0,
+				versions: [],
+			};
+		}
+
+		packagesVersions[name].usages += 1;
+		if (!packagesVersions[name].versions.includes(version)) {
+			packagesVersions[name].versions.push(version);
+		}
+	}
+};
+
+export async function workflow({ files, dirs, question, exec }: Api) {
+	const workspaceFile = files("pnpm-workspace.yaml").yaml();
+	const workspaceConfig = (
+		await workspaceFile.map(({ getContents }) =>
+			getContents<{ packages: string[] }>(),
+		)
+	).pop();
+
+	if (!workspaceConfig) {
+		console.log("pnpm-workspace.yaml not found");
+		return;
+	}
+
+	const packagesVersions: PackagesVersions = {};
+
+	const packageJsonFiles = dirs({
+		dirs: workspaceConfig.packages,
+		ignore: ["**/node_modules/**"],
+	}).files("package.json");
+
+	await packageJsonFiles.json(async ({ map }) => {
+		const packageJson = (
+			await map(({ getContents }) => getContents<PackageJson>())
+		).pop();
+
+		if (!packageJson) {
+			return;
+		}
+
+		readDependencies(packagesVersions, packageJson.dependencies);
+		readDependencies(packagesVersions, packageJson.devDependencies);
+	});
+
+	if (!Object.entries(packagesVersions).some(([_, { usages }]) => usages > 1)) {
+		console.log("No duplicated dependencies found");
+		return;
+	}
```

Comment:
So we replace only duplicated items? I think it would make sense to move all versions to catalogs. If in the future a new package will depend on it, the version will already be in the catalog.

Reply:
sounds good, will do that